app/jobs/identify_external_clients_job.rb

```diff
@@ -0,0 +1,152 @@
+        end
+
+        log("matched #{output_rows.size} of #{input_rows.size} rows", type: :info, object_key: input_key)
+
+        # s3.store raises on failure
+        outbox_s3.store(
+          content: format_as_csv(output_rows.compact),
+          name: upload_name(input_key),
+          content_type: 'text/csv; charset=utf-8',
+        )
+
+        # if we successfully processed the input object, delete it from the bucket
+        inbox_s3.delete(key: input_key)
+      end
+    end
+  end
+
+  protected
+
+  def upload_name(filename)
+    dir = File.dirname(filename)
+    dir = dir == '.' ? nil : dir
+    base = File.basename(filename, '.*')
+    ext = File.extname(filename)
+    [dir, "#{base}-results#{ext}"].compact.join('/')
+  end
+
+  private def build_lookups
+    @name_lookup = GrdaWarehouse::ClientMatcherLookups::ProperNameLookup.new(transliterate: true)
+    @ssn_lookup = GrdaWarehouse::ClientMatcherLookups::SSNLookup.new(format: :last_four)
+    @dob_lookup = GrdaWarehouse::ClientMatcherLookups::DOBLookup.new
+
+    clients = GrdaWarehouse::Hud::Client.joins(:warehouse_client_source).source
+    wc_t = GrdaWarehouse::WarehouseClient.arel_table
+    id_field = Arel.sql(wc_t[:destination_id].to_sql)
+    GrdaWarehouse::ClientMatcherLookups::ClientStub.from_scope(clients, id_field: id_field) do |client|
+      @name_lookup.add(client)
+      @ssn_lookup.add(client)
+      @dob_lookup.add(client)
+    end
+  end
+
+  def with_lock(&block)
+    GrdaWarehouseBase.with_advisory_lock('identify_external_clients', timeout_seconds: 0, &block)
+  end
+
+  def parse_csv_string(csv_string, key:)
+    results = []
+    begin
+      csv = CSV.parse(csv_string, headers: true)
+      results = csv.map do |row|
+        # normalize headers
+        csv.headers.each_with_object({}) do |header, hash|
+          hash[header.downcase] = row[header]
+        end
+      end
+    rescue CSV::MalformedCSVError => e
+      log("CSV parsing error: #{e.message}", object_key: key)
+    rescue ArgumentError => e
+      log("Argument error (possibly encoding related): #{e.message}", object_key: key)
+    rescue StandardError => e
+      log("An unexpected error occurred: #{e.message}", object_key: key)
+    end
+    results
+  end
+
+  def process_row(row, external_id_field)
+    first_name, last_name, ssn4, dob, external_id = row.values_at('first_name', 'last_name', 'ssn4', 'dob', external_id_field.downcase)
+    return if external_id.blank?
+
+    matches = [
+      @ssn_lookup.get_ids(ssn: ssn4),
+      @dob_lookup.get_ids(dob: dob),
+      @name_lookup.get_ids(first_name: first_name, last_name: last_name),
+    ]
+
+    match = get_best_match(matches)
+    match ? { external_id_field => external_id, 'Client ID' => match } : nil
+  end
+
+  # matching id is found in at least two match sets
+  def get_best_match(match_sets, min_rank: 2)
+    ranks = Hash.new(0)
+    match_sets.each do |matches|
+      matches.each { |match| ranks[match] += 1 }
+    end
```

Comment:
Pretty sure this is the same as:
```suggestion
    ranks =  match_sets.flatten.tally
```


```
irb(main):004:0> match_sets = [[1,2,3], [2, 3], [3]]
=> [[1, 2, 3], [2, 3], [3]]
irb(main):005:0> ranks = Hash.new(0)
=> {}
irb(main):006:1* match_sets.each do |matches|
irb(main):007:1*        matches.each { |match| ranks[match] += 1 }
irb(main):008:0> end
=> [[1, 2, 3], [2, 3], [3]]
irb(main):009:0> ranks
=> {1=>1, 2=>2, 3=>3}
irb(main):010:0> match_sets.flatten.tally
=> {1=>1, 2=>2, 3=>3}
irb(main):011:0>
```